### PR TITLE
WIP: Make shape a pointer and add ndim + strides

### PR DESCRIPTION
**Pull request by jakirkham**
Adds `_shape` and `_strides` as private pointers that contain the shape and strides of the data. Also adds `ndim` to aid tracking the dimensionality as a publicly visible value. Includes properties to view the `shape` and `strides`.

Note: Still need to handle some code that is assigning to `self.shape` as that is no longer valid with this approach.

### CODE REVIEW

**Comment by mrocklin** [comment]
Right, I get this error 

```python-traceback
Traceback (most recent call last):
  File "/home/nfs/mrocklin/miniconda/envs/ucx/lib/python3.7/site-packages/tornado/gen.py", line 736, in run
    yielded = self.gen.throw(*exc_info)  # type: ignore
  File "/home/nfs/mrocklin/distributed/distributed/nanny.py", line 354, in _on_exit
    yield self.scheduler.unregister(address=self.worker_address)
  File "/home/nfs/mrocklin/miniconda/envs/ucx/lib/python3.7/site-packages/tornado/gen.py", line 729, in run
    value = future.result()
  File "/home/nfs/mrocklin/miniconda/envs/ucx/lib/python3.7/site-packages/tornado/gen.py", line 736, in run
    yielded = self.gen.throw(*exc_info)  # type: ignore
  File "/home/nfs/mrocklin/distributed/distributed/core.py", line 741, in send_recv_from_rpc
    result = yield send_recv(comm=comm, op=key, **kwargs)
  File "/home/nfs/mrocklin/miniconda/envs/ucx/lib/python3.7/site-packages/tornado/gen.py", line 729, in run
    value = future.result()
  File "/home/nfs/mrocklin/miniconda/envs/ucx/lib/python3.7/site-packages/tornado/gen.py", line 736, in run
    yielded = self.gen.throw(*exc_info)  # type: ignore
  File "/home/nfs/mrocklin/distributed/distributed/core.py", line 533, in send_recv
    yield comm.write(msg, serializers=serializers, on_error="raise")
  File "/home/nfs/mrocklin/miniconda/envs/ucx/lib/python3.7/site-packages/tornado/gen.py", line 729, in run
    value = future.result()
  File "/home/nfs/mrocklin/distributed/distributed/comm/ucx.py", line 113, in write
    await self.ep.send_obj(meta)
  File "ucp/_libs/ucp_py.pyx", line 370, in ucp_py.Endpoint.send_obj
  File "ucp/_libs/ucp_py.pyx", line 348, in ucp_py.Endpoint._send_obj_host
  File "ucp/_libs/ucp_py_buffer_helper.pyx", line 204, in ucp_py.BufferRegion.populate_ptr
  File "ucp/_libs/ucp_py_buffer_helper.pyx", line 207, in ucp_py.BufferRegion._populate_ptr
AttributeError: attribute 'shape' of 'ucp_py.BufferRegion' objects is not writable
ERROR    asyncio:base_events.py:1608 Future exception was never retrieved
future: <Future finished exception=AttributeError("attribute 'shape' of 'ucp_py.BufferRegion' objects is not writable")>
```

when I run this test (on latest distributed master)

```
py.test distributed/comm/tests/test_ucx.py::test_stress -s --runslow
```

**Comment by jakirkham** [comment]
Yep, that's expected currently.

One of the things I'm trying to wrap my head around ATM is what kinds of buffers are we trying to handle here?

For instance [in `master` `ndim = 1`]( https://github.com/rapidsai/ucx-py/blob/4030eff63438ac87d58fd009656198a184914041/ucp/_libs/ucp_py_buffer_helper.pyx#L164 ). Is the general idea to only support 1-D buffers in this class? If so, the code can probably be simplified a lot and this assignment issue dealt with much more simply.

Also are we expecting buffers to be contiguous or may they be non-contiguous?

**Comment by jakirkham** [comment]
Perhaps these are questions for @TomAugspurger or @Akshay-Venkatesh?

**Comment by TomAugspurger** [comment]
I thought that everything was 1D by this point, but that may have been
mistaken?

**Comment by mrocklin** [comment]
The fact that my PR fixes a failure implies that that might not be true.

**Comment by jakirkham** [comment]
Makes sense. I'm happy to keep pushing on this then. Just want to make sure I'm not overcomplicating things.

If anyone has thoughts on whether data is required to be contiguous or not, that would be good to know.

**Comment by mrocklin** [comment]
I don't know what the implementation does.  I recall conversations where we said "lets not bother with non-contiguous data.  If it's non-contiguous, let's just make a copy".  I can't authoritatively say that that's what happened though.

**Comment by TomAugspurger** [comment]
That's my recollection as well.

**Comment by jakirkham** [comment]
Is there a way to try sending a NumPy or CuPy array through UCX and seeing what you get out? I think this would be really informative as we could quickly see how these different cases are handled and get several test cases to include as a bonus.

**Comment by Akshay-Venkatesh** [comment]
@jakirkham [This](https://github.com/rapidsai/ucx-py/blob/devel/tests/test_send_recv_obj.py) test written by @TomAugspurger is probably the easiest way to check what is being received.

**Comment by mrocklin** [comment]
When I try this with the ucx stress test I get the following error:

```python-traceback
  File "/home/nfs/mrocklin/distributed/distributed/worker.py", line 742, in _register_with_scheduler
    serializers=["msgpack"],
  File "/home/nfs/mrocklin/miniconda/envs/ucx/lib/python3.7/site-packages/tornado/gen.py", line 729, in run
    value = future.result()
  File "/home/nfs/mrocklin/distributed/distributed/comm/ucx.py", line 113, in write
    await self.ep.send_obj(meta)
  File "ucp/_libs/ucp_py.pyx", line 370, in ucp_py.Endpoint.send_obj
  File "ucp/_libs/ucp_py.pyx", line 348, in ucp_py.Endpoint._send_obj_host
  File "ucp/_libs/ucp_py_buffer_helper.pyx", line 211, in ucp_py.BufferRegion.populate_ptr
  File "ucp/_libs/ucp_py_buffer_helper.pyx", line 214, in ucp_py.BufferRegion._populate_ptr
AttributeError: attribute 'shape' of 'ucp_py.BufferRegion' objects is not writable
```

I think that the proper thing to do here is to try this with the following test:

```
py.test distributed/comm/tests/test_ucx.py::test_stress --runslow
```

**Comment by jakirkham** [comment]
@mrocklin, I haven't been doing anything on this since we last discussed it. The commit pushed above included some changes locally that needed to be moved out of my local copy before doing other work. I doubt they (or this PR more generally) are complete as the commit message notes. Happy to take a look at this again if you would like. 🙂

**Comment by jakirkham** [comment]
Independently we don't seem to be handling buffer protocol requests correctly. Have raised issue ( https://github.com/rapidsai/ucx-py/issues/127 ) to discuss that.

**Comment by mrocklin** [comment]
@madsbk could I ask you to take a look at this PR and https://github.com/rapidsai/ucx-py/pull/120 ?

File: ucp/_libs/ucp_py_buffer_helper.pyx [hunk 1]
```diff
@@ -107,18 +110,43 @@ cdef class BufferRegion:
...
         self._readonly = False  # True?
         self.buf = NULL
-        self.shape = [0]
+        self.ndim = 1
+        self._shape = <Py_ssize_t*>malloc(sizeof(Py_ssize_t) * self.ndim)
+        for i in range(self.ndim):
+            self._shape[i] = 0
+        self._strides = <Py_ssize_t*>malloc(sizeof(Py_ssize_t) * self.ndim)
+        for i in range(self.ndim):
+            self._strides[i] = 1
```
Comment (madsbk) [inline ucp/_libs/ucp_py_buffer_helper.pyx]: To support multi-dimension `memoryviews`, you could something like:
```python
    s = 1
    for i in reversed(range(self.ndim)):
        self._strides[i] = s
        s *= self._strides[i]
```

File: ucp/_libs/ucp_py_buffer_helper.pyx [hunk 2]
```diff
@@ -107,18 +110,43 @@ cdef class BufferRegion:
...
+        free(self._strides)
+        self._strides = <Py_ssize_t*>malloc(sizeof(Py_ssize_t) * self.ndim)
+        self._strides[0] = 1
 
     def __len__(self):
         if not self.is_set:
             return 0
         else:
-            return self.shape[0]
+            return self._shape[0]
```
Comment (madsbk) [inline ucp/_libs/ucp_py_buffer_helper.pyx]: Ditto
```python
        size = 1
        for i in range(self.ndim):
            size *= self._shape[i]
```

File: ucp/_libs/ucp_py_buffer_helper.pyx [hunk 3]
```diff
@@ -141,31 +169,25 @@ cdef class BufferRegion:
...
-            Py_ssize_t shape2[1]
             empty = b''
 
         if not self.is_set:
             raise ValueError("This buffer region's memory has not been set.")
 
-        strides[0] = <Py_ssize_t>self.itemsize
-        assert len(self.shape)
-        if self.shape[0] == 0:
+        if self._shape[0] == 0:
```
Comment (madsbk) [inline ucp/_libs/ucp_py_buffer_helper.pyx]: Ditto
```python
if len(self) == 0:
```

File: ucp/_libs/ucp_py_buffer_helper.pyx [hunk 4]
```diff
@@ -141,31 +169,25 @@ cdef class BufferRegion:
...
             buffer.buf = <void *>&(self.buf.buf[0])
 
-        shape2[0] = self.shape[0]
-
         buffer.format = self.format
         buffer.internal = NULL
         buffer.itemsize = self.itemsize
-        buffer.len = self.shape[0] * self.itemsize
-        buffer.ndim = 1
+        buffer.len = self._shape[0] * self.itemsize
```
Comment (madsbk) [inline ucp/_libs/ucp_py_buffer_helper.pyx]: Ditto
```python
buffer.len = self.nbytes()
```

File: ucp/_libs/ucp_py_buffer_helper.pyx (hunk 1 above)
Comment (madsbk) [reply ucp/_libs/ucp_py_buffer_helper.pyx]: Are `BufferRegion.strides()` supposed to be in elements or bytes?

**Comment by jakirkham** [comment]
Feel free to push changes directly to this PR @madsbk if you are comfortable. 🙂

**Comment by madsbk** [comment]
OK but I just realized that `populate_ptr()` and `populate_cuda_ptr()` also needs to support `ndim` change. 
@jakirkham, is it alright with you if I create a new PR that rewrites `BufferRegion` into a more clean implementation that incorporate your ideas?

**Comment by jakirkham** [comment]
Please go ahead, @madsbk. 🙂

**Comment by jakirkham** [comment]
I think this was basically addressed by @madsbk's PR ( https://github.com/rapidsai/ucx-py/pull/144 ). So closing it out. Though feel free to pull things from this if it winds up still being useful for some reason.

**Pull request closed without merging**